File: src/main/java/com/example/message/mapper/IMapper.java
package com.example.message.mapper;

import java.util.List;
import java.util.stream.Collectors;

public interface IMapper<Domain, Dto> {

    Domain mapToDomain(Dto dto);
    Dto mapToDto(Domain domain);

    default List<Domain> mapToDomainList(List<Dto> dtoList){
        return dtoList.stream().map(this::mapToDomain).collect(Collectors.toList());
    }

    default List<Dto> mapToDtoList(List<Domain> domainList){
        return domainList.stream().map(this::mapToDto).collect(Collectors.toList());
    }
}

File: src/main/java/com/example/message/mapper/IMapper.kt
package com.example.message.mapper

interface IMapper<Domain, Dto> {

    fun mapToDomain(dto: Dto): Domain
    fun mapToDto(domain: Domain): Dto

    fun mapToDomainList(dtoList: List<Dto>): List<Domain> {
        return dtoList.map { dto: Dto -> mapToDomain(dto) }
    }

    fun mapToDtoList(domainList: List<Domain>): List<Dto> {
        return domainList.map { domain: Domain -> mapToDto(domain) }
    }
}